### ISSUE TITLE
Add support for Net 9

### ISSUE THREAD

**Issue by rizi** [JasperFx/lamar#405]
I opend a [PR](https://github.com/JasperFx/lamar/pull/404) to update lamar to net 9 and all of its dependencies.

However there seems to be problem with aspnet.core 9 and lamar, I provided a detailed description in the PR.

**This error ocurrs when an aspnet.core 9 application is started:**
Unable to resolve service for type '**System.Collections.Generic.IEnumerable`1[Microsoft.AspNetCore.Diagnostics.IDeveloperPageExceptionFilter]**' while attempting to activate 'Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddlewareImpl'.
at Microsoft.Extensions.DependencyInjection.ActivatorUtilities.ConstructorMatcher.CreateInstance(IServiceProvider provider)
at Microsoft.Extensions.DependencyInjection.ActivatorUtilities.CreateInstance(IServiceProvider provider, Type instanceType, Object[] parameters)

Any help would highly be appreciated.

**Comment by miholler** [comment]
I ran into the same issue and I would like to share some of my observations. Maybe they are helpful to track down the bug.

The error happens within the ResolveFamily method of the ServiceGraph class. During startup ASP.Net is calling the Scope->GetService method by passing in IEnumerable of IDeveloperPageExceptionFilter as serviceType which is ultimately passed to the ResolveFamily method. For some reason the call to _families.TryFind(serviceType, out var value) within ResolveFamily returns true and finds a family, but this family is defined for the IDeveloperExceptionPageFilter type instead of the collection type. This leads to GetService returning null instead of an empty list and ASP.Net is throwing the mentioned exception.

When doing the call to GetService manually, the family is not found and addMissingFamily is called within ResolveFamily which leads to creating a family for the list type and returning an empty list. When ASP.Net is calling for the list afterwards, the correct family is pulled from _families, an empty list is passed to the DeveloperExceptionPageMiddlewareImpl class and the application starts.

**Comment by rizi** [comment]
> I ran into the same issue and I would like to share some of my observations. Maybe they are helpful to track down the bug.
> 
> The error happens within the ResolveFamily method of the ServiceGraph class. During startup ASP.Net is calling the Scope->GetService method by passing in IEnumerable of IDeveloperPageExceptionFilter as serviceType which is ultimately passed to the ResolveFamily method. For some reason the call to _families.TryFind(serviceType, out var value) within ResolveFamily returns true and finds a family, but this family is defined for the IDeveloperExceptionPageFilter type instead of the collection type. This leads to GetService returning null instead of an empty list and ASP.Net is throwing the mentioned exception.
> 
> When doing the call to GetService manually, the family is not found and addMissingFamily is called within ResolveFamily which leads to creating a family for the list type and returning an empty list. When ASP.Net is calling for the list afterwards, the correct family is pulled from _families, an empty list is passed to the DeveloperExceptionPageMiddlewareImpl class and the application starts.

@miholler 
Thx for the info, I found the reason (exactly as you described it) , O also found a fix, but I have to find a better solution.

I also talked to @jeremydmiller via discord, I will try to find a suitable fix, as soon as I got time.

Br

**Comment by miholler** [comment]
@rizi thx for the quick reply. I came up with a small workaround, which I will remove as soon as your bugfix becomes available, but I would like to share it for anyone having the same problem at the moment:
`builder.Host.UseLamar((ctx, registry) => registry.ForSingletonOf<IEnumerable<IDeveloperPageExceptionFilter>>().UseIfNone([]));`

**Comment by rizi** [comment]
@miholler I fixed the issue and updated the PR accordingly.

I had to partially revert the change from #391.

From now on we have to use the [FromBody] attribute when a controller method or a minimal api gets passed an IEnumerable{T} parameter, all other collection types are still supported (e.g. IReadOnlyCollection{T}, IList{T}, IReadOnlyList{T}, List{T}, ...) and don't need the [FromBody] attribute.

This change is necessary because MS used IEnumerable{IDeveloperPageExceptionFilter} in a middlware (of course this would also be necessary when MS would have used IEnumerable{T} somewhere else), so Lamar must handle such use cases.

**For example**

Still works as before:

```cs
public async Task UpdateAsync(IReadOnlyCollection<UpdateCommand> updateCommands) {} 
```

For IEnumerable{T} we have to use the [FromBody] attribute

```cs
public async Task UpdateAsync([FromBody] IEnumerable<UpdateCommand> updateCommands) {} 
```
